lighthouse-core/test/gather/gatherers/js-usage-test.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env jest */
+
+jest.useFakeTimers();
+
+const Driver = require('../../../gather/driver.js');
+const Connection = require('../../../gather/connections/connection.js');
+const JsUsage = require('../../../gather/gatherers/js-usage.js');
+const {createMockSendCommandFn, createMockOnFn} = require('../mock-commands.js');
+
+describe('JsUsage gatherer', () => {
+  /**
+   * `scriptParsedEvent` mocks the `sourceMapURL` and `url` seen from the protocol.
+   * `map` mocks the (JSON) of the source maps that `Runtime.evaluate` returns.
+   * `resolvedSourceMapUrl` is used to assert that the SourceMaps gatherer is using the expected
+   *                        url to fetch the source map.
+   * `fetchError` mocks an error that happens in the page. Only fetch error message make sense.
+   * @param {{coverage: LH.Crdp.Profiler.ScriptCoverage[], scriptParsedEvents: LH.Crdp.Debugger.ScriptParsedEvent[]}} _
+   * @return {Promise<LH.Artifacts['JsUsage']>}
+   */
+  async function runJsUsage({coverage, scriptParsedEvents = []}) {
+    const onMock = createMockOnFn();
+    const sendCommandMock = createMockSendCommandFn()
+      .mockResponse('Profiler.enable', {})
+      .mockResponse('Profiler.disable', {})
+      .mockResponse('Profiler.startPreciseCoverage', {})
+      .mockResponse('Profiler.takePreciseCoverage', {result: coverage})
+      .mockResponse('Profiler.stopPreciseCoverage', {});
+
+    for (const scriptParsedEvent of scriptParsedEvents) {
+      onMock.mockEvent('protocolevent', {
+        method: 'Debugger.scriptParsed',
+        params: scriptParsedEvent,
+      });
+    }
+    const connectionStub = new Connection();
+    connectionStub.sendCommand = sendCommandMock;
+    connectionStub.on = onMock;
+
+    const driver = new Driver(connectionStub);
+
+    const gatherer = new JsUsage();
+    await gatherer.beforePass({driver});
+    jest.advanceTimersByTime(1);
```

Comment:
where are the timeouts coming from?

Reply:
event callback doesnt happen in time without this

Reply:
ah the script parsed listeners! comment that? :)